[API] Sub devices and areas

dala318 opened the pull request:
# What does this implement/fix?

This PR adds support for sub-devices and areas in ESPHome, enabling a single ESP to represent multiple logical devices in Home Assistant. This is a foundational feature for building smart home hubs and gateways.

Additionally, this PR adds validation to prevent duplicate entity names within the same platform and device, providing clearer configuration errors.

## Primary Use Cases

### 1. Hub/Gateway Functionality
When an ESP acts as a bridge for multiple devices connected through various protocols:
- **RF bridges**: Each 433MHz/868MHz device (door sensors, PIR sensors, etc.) appears as its own device in Home Assistant rather than entities under the bridge
- **Modbus gateways**: Individual Modbus devices (energy meters, HVAC controllers) get their own device representation
- **I2C/SPI sensor arrays**: Multiple sensors on the same bus can be grouped as separate logical devices
- **Serial/UART bridges**: Devices connected via serial protocols can each have their own identity

### 2. Multi-Zone Controllers
- One ESP controlling multiple rooms or areas
- Each zone appears as a separate device with its own area assignment
- Better organization without deploying multiple physical ESPs

## Key Features

### Sub-Devices and Areas
- Group entities into logical sub-devices that appear separately in Home Assistant
- Assign sub-devices to specific areas (rooms) for better organization
- Each sub-device can have its own area assignment independent of the ESP's physical location

### Flexible Configuration
- Define areas using either simple string format or structured format with `id` and `name`
- Create multiple sub-devices, each assignable to different areas
- Entities can be assigned to specific sub-devices using `device_id`

### Implementation Details
- Uses hash-based unique IDs for devices and areas
- Memory-efficient design using `const char*` for names
- Separate compile-time flags (`USE_AREAS` and `USE_DEVICES`) to minimize overhead
- Areas can be used without devices if needed
- Validates entity names to prevent duplicates within the same platform
- Duplicate detection based on (platform, sanitized_name) tuple
- Same entity names only work across different platforms (not across devices due to hash collisions)

### API Protocol
- New `AreaInfo` and `DeviceInfo` messages in API protocol
- Entities report their device association to Home Assistant
- Clean separation between physical ESP device and sub-devices

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**
- fixes https://github.com/esphome/feature-requests/issues/1335
- fixes https://github.com/esphome/issues/issues/6953

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
- esphome/esphome-docs#4813

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

### Example 1: RF Bridge Gateway
```yaml
# ESP32 acting as RF bridge for multiple 433MHz devices
esphome:
  name: rf-bridge
  area:
    id: bridge_area
    name: "Utility Room"
  
  devices:
    - id: front_door_device
      name: "Front Door Sensor"
      area_id: entrance_area
    - id: kitchen_motion_device
      name: "Kitchen Motion"
      area_id: kitchen_area
    - id: garage_door_device
      name: "Garage Door"
      area_id: garage_area
  
  areas:
    - id: entrance_area
      name: "Entrance"
    - id: kitchen_area
      name: "Kitchen"
    - id: garage_area
      name: "Garage"

# RF receiver configuration
remote_receiver:
  pin: GPIO4
  dump: rc_switch

# Each RF device appears as a separate device in HA
binary_sensor:
  - platform: remote_receiver
    name: "Front Door"
    device_id: front_door_device
    rc_switch_raw:
      code: '101010110101'
      
  - platform: remote_receiver
    name: "Kitchen Motion"
    device_id: kitchen_motion_device
    rc_switch_raw:
      code: '110011001100'
      
  - platform: remote_receiver
    name: "Garage Door"
    device_id: garage_door_device
    rc_switch_raw:
      code: '111000111000'
```

### Example 2: Multi-Zone Controller
```yaml
# ESP32 controlling multiple rooms
esphome:
  name: multi-room-controller
  
  devices:
    - id: living_room_device
      name: "Living Room Controller"
      area_id: living_area
    - id: kitchen_device
      name: "Kitchen Controller"
      area_id: kitchen_area
  
  areas:
    - id: living_area
      name: "Living Room"
    - id: kitchen_area
      name: "Kitchen"

sensor:
  - platform: dht
    pin: GPIO5
    temperature:
      name: "Temperature"
      device_id: living_room_device
    humidity:
      name: "Humidity"
      device_id: living_room_device

switch:
  - platform: gpio
    pin: GPIO10
    name: "Light"
    device_id: living_room_device
```

### Example 3: Area Configuration
```yaml
# Area configuration for the ESP device
esphome:
  name: my-device
  area:
    id: living_room
    name: "Living Room"

# Additional areas can be defined for sub-devices
  areas:
    - id: kitchen
      name: "Kitchen"
    - id: bedroom
      name: "Bedroom"

# Sub-devices can reference these areas
  devices:
    - id: kitchen_controller
      name: "Kitchen Controller"
      area_id: kitchen
```

### Example 4: Entity Name Validation
```yaml
# Duplicate entity names on same device/platform will fail validation
binary_sensor:
  - platform: template
    name: "Motion"
    lambda: return false;
    
  - platform: template
    name: "Motion"  # ERROR: Duplicate binary_sensor entity with name 'Motion' found
    lambda: return true;

# Duplicates across different platforms are allowed
sensor:
  - platform: template
    name: "Motion"  # OK - different platform
    lambda: return 0.0;

# Duplicates on different devices don't work (discovered during testing)
# This causes hash collisions in Home Assistant
binary_sensor:
  - platform: template
    name: "Status"
    device_id: device_1
    lambda: return false;
    
  - platform: template
    name: "Status"  # ERROR: Hash collision - won't work in HA
    device_id: device_2
    lambda: return true;
```

## Area Configuration Formats

ESPHome supports two formats for configuring areas:

### Simple String Format
```yaml
# Simple string format (still supported)
esphome:
  name: my-device
  area: "Kitchen"
```

### Structured Format
```yaml
# Structured format (recommended when using sub-devices)
esphome:
  name: my-device
  area:
    id: kitchen
    name: "Kitchen"
```

**Note**: Both formats are supported. The simple string format remains valid for basic use cases where you just want to assign the ESP device to a room. The structured format with `id` and `name` fields is recommended when you plan to use sub-devices, as the `id` field allows you to reference areas when creating sub-devices (using `area_id` in device configuration).

## Breaking Changes

### Duplicate Entity Names Now Fail Validation

Previously, ESPHome would silently accept duplicate entity names within the same platform. This caused problems in Home Assistant where duplicate entities would not appear correctly and generate warnings about duplicate unique IDs.

**What's changed:**
- Duplicate entity names within the same platform now fail validation with a clear error message, even if they're on different devices
- The error tells you exactly which entity is duplicated
- This prevents the confusing situation where entities work in ESPHome but fail to appear correctly in Home Assistant

**Important Note:** During testing, it was discovered that entities with the same name on different devices cause hash collisions in Home Assistant, making them indistinguishable. While we could potentially add device_id to the EntityState messages to support this in the future, such a change would require weeks of planning and execution to avoid breaking existing Home Assistant installations. For now, we enforce unique names across all devices to prevent these issues.

**What you need to do:**
If your configuration has duplicate entity names, you'll see an error like:
```
Duplicate binary_sensor entity with name 'Motion' found. Each entity must have a unique name within its platform across all devices.
```

To fix this, give each entity a unique name:
```yaml
# Before (would previously be accepted, now fails):
binary_sensor:
  - platform: gpio
    name: "Motion"
    pin: GPIO5
  - platform: gpio
    name: "Motion"  # ERROR: Duplicate name
    pin: GPIO6

# After (fixed):
binary_sensor:
  - platform: gpio
    name: "Front Motion"
    pin: GPIO5
  - platform: gpio
    name: "Back Motion"
    pin: GPIO6
```

**Note:** This applies to all entities within the same platform, regardless of device. You can still have:
- Same names across different platforms (e.g., a `sensor.temperature` and `binary_sensor.temperature`)

What we discovered during testing:
- Same names on different devices within the same platform don't work due to hash collisions in Home Assistant
- While this might have appeared to work in ESPHome, it would fail in Home Assistant with duplicate ID errors
- We now detect and prevent this at validation time to avoid confusion

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

dala318 committed:
Add device_id to entity_base

dala318 committed:
Add a device component

dala318 committed:
Some basic chain working

dala318 committed:
Add a single test

dala318 committed:
Add code-owner tag

dala318 committed:
Formating updates

dala318 committed:
Register device_id to entity and separate struct for all device info

dala318 committed:
Fix generated proto-files

dala318 committed:
Add CODEOWNER

dala318 committed:
Store proto with all additions but commented out

dala318 committed:
Work on sub-device creation

dala318 committed:
Fix clang-tidy

dala318 committed:
Make it a Class but only use the id in entities

dala318 committed:
Cleaning up

dala318 committed:
Fix clang

nielsnl68 commented:
Interesting idea. I will follow your progress.

dala318 committed:
Merge branch 'dev' into multi_device

dala318 committed:
Improve stability for unrelated test

dala318 committed:
Merge branch 'dev' into multi_device

dala318 commented:
Marking as ready for review even if for now only 2 entity types are implemented, to make change properly visible for core maintainers. Once the overall concept is agreed/accepted I can bring it back to draft and continue with the rest of entity types.

dala318 committed:
Merge branch 'dev' into multi_device

dala318 committed:
Fix generated code after merge

dala318 committed:
Activate the rest of entities

dala318 committed:
Merge pull request #10 from dala318/multi_device_2

Sub Devices (all entities)

dala318 committed:
Rename ref in codegen

dala318 committed:
Merge branch 'dev' into multi_device

bdraco commented:
One thing to keep in mind is that this does add some complexity in HA, since we’ll need to manage multiple devices — including handling add/remove/update when changes happen. We’ll also need to figure out how to model the parent/child relationships. That might not be too bad though, especially if all the sub-devices can just be tied to the main device.

bdraco commented:
> @bdraco would this also help with the ideas i opted last month with you? I was thinking it is.

You'll have to refresh my memory as that was 5000 context switches ago :)

dougiteixeira commented:
I have some considerations, just to contribute to the discussion:

* Adding subdevices: created when the entity is created/updated, as happens with most integrations.

* Removing devices and subdevices:
a) The main device and subdevices should be removed when deleting the configuration entry.
b) If you want to delete a subdevice (perhaps because it was removed from the code/firmware), you can implement the manual removal on the device page (something already common in Home Assistant), I think it is acceptable due to the DIY nature of ESPHome.

* The parent/child model, I believe it should be very simple: the parent device is the physical device and the subdevices are linked to the parent (via_device).

Although it adds an extra layer of complexity to the ESPHome code in Home Assistant, the way it is being implemented is simple for the end user and will bring versatility and organization to projects with ESPHome.

dala318 commented:
There might be a need for two different implementations for solving what I understand as two different needs.

If I understand the ESPNOW related comment above correctly, that aims to identify another **ESP-device** that connects to HA via the ESPHome device as a device of its own. That to me more sounds like how Bluetooth devices that are connected vi the Bluetooth Proxy are identified in HA as connected via an ESPHome device. (How those can "jump" between different proxies in HA I'm not aware of). Those **devices** have a full status and identifier.

This change aims to group a selected group of **entities** in the ESPHome config as a device of its own. Handy if say one ESP serves several purposes. Physically one device, but logically more than one usage, ex. Temp and pressure for water circuit + Relay for fence gate. Such **sub-devices** have only the states and function related to the selected entities. The **main device** will be the one that reports the actual device status and get OTA updates etc.

nielsnl68 commented:
@dala318, you did understand me correctly.
And i understand your approach as well and like you see this as a good starting point.
As long you can make as open as possible to hook other ideas into the same infrastructure i'm already happy.
At a later point we can think of a way to implement something like the bluetooth proxy. 

But for me your work is essential for what i have in mind.

dala318 commented:
> As long you can make as open as possible to hook other ideas into the same infrastructure i'm already happy.
> At a later point we can think of a way to implement something like the bluetooth proxy. 

I'm all for open an expandable implementation. So if you have ideas what might be needed for making this more compatible to your needs please make suggestions.

jesserockz commented on .github/workflows/ci-api-proto.yml:
New PR if you want this in CI please

dala318 replied:
Oh, yes agree.
I added it only to get the expected result of the codegen since I never got the environment properly set-up locally to generate the files. Planning to remove it once the rest is approved.
Maybe a more proper addition in another PR would be to archive the generated artifacts so that one could download and commit them, avoiding the hazzle to copy-paste git-diffs.

dala318 replied:
Created https://github.com/esphome/esphome/pull/8698

jesserockz commented on esphome/components/api/api.proto:
```diff
@@ -186,6 +186,12 @@ message DeviceInfoRequest {
 }
 
+message SubDeviceInfo {
+  string id = 1;
```
Do we need a string here?
Can we get away with a `fixed32 id` and do all the linking in the code?

dala318 replied:
Yes, seem like a more robust solution.

dala318 replied:
Changed into uint32

jesserockz commented on esphome/components/devices/__init__.py:
I feel like this should go into `core/config.py` to live under the `esphome` key in configuration and not be a component

dala318 replied:
To do this I think I would need some more guidance from the more experienced participants.
I added it as a separate component to make it fully optional and easier to scale and adapt in the future. (And not interfere to much with the core config.)
What are the benefits or rationale of fusing into the `esphome` config part?
That part holds the "main" device information, and yes, this is a partial clone to that and the sub-device data goes into the same API data as most of what's in that config-part.
I really don't have any objections to moving this to `core/config.py`, just want to make sure it's the agreed way forward.

dala318 replied:
I've made an attempt in removing the `components/device/` stuff and put it in `core/config.py` and a new `core/sub_devices.h`
Please review if it's how you expected.

jesserockz commented on esphome/core/entity_base.h:
```diff
@@ -61,6 +65,7 @@ class EntityBase {
   bool internal_{false};
   bool disabled_by_default_{false};
   EntityCategory entity_category_{ENTITY_CATEGORY_NONE};
+  StringRef device_id_{""};
```
This really should just be a pointer to the SubDevice object

dala318 replied:
Is now an uint32_t

jesserockz commented on tests/dashboard/test_web_server.py:
Separate PR please

dala318 replied:
I was hoping to revert this part, got some intermittent test crashes in CI.
Can try to remove, if it keep happening I will push it as a separate PR

dala318 replied:
Added PR https://github.com/esphome/esphome/pull/8699
Change reverted on this branch

nielsnl68 commented:
> > As long you can make as open as possible to hook other ideas into the same infrastructure i'm already happy.
> > At a later point we can think of a way to implement something like the bluetooth proxy.
> 
> I'm all for open an expandable implementation. So if you have ideas what might be needed for making this more compatible to your needs please make suggestions.

For now, please make it bidirectional. And i think everything is fine.

dala318 committed:
Merge branch 'dev' into multi_device

dala318 commented on esphome/components/api/api_connection.cpp:
```diff
@@ -337,6 +338,7 @@ bool APIConnection::try_send_cover_info(APIConnection *api, void *v_cover) {
   msg.entity_category = static_cast<enums::EntityCategory>(cover->get_entity_category());
+  msg.device_id = cover->get_device_id();
   return api->send_list_entities_cover_response(msg);
 }
```
Seem to have forgotten a whole bunch of entity types, will fix.

dala318 committed:
Attempt to replace device_id:str with device_uid:uint32

dala318 committed:
Fix generated files

dala318 committed:
Fix generated files and revert entity config to device_id

dala318 committed:
Fix clang

dala318 committed:
Restore ci-api-proto.yml

dala318 committed:
Revert "Improve stability for unrelated test"

This reverts commit 3922950951191ed3052964b000dac2651595c419.

dala318 committed:
Attempt moving it to esphome config section

dala318 committed:
remove from CODEOWNERS

dala318 committed:
Fix namespace error

dala318 commented on esphome/config_validation.py:
```diff
@@ -347,6 +348,13 @@ def icon(value):
 
+def sub_device_id(value):
+    # Duplicate definition of SubDevice to avoid circular import
+    SubDevice = cg.esphome_ns.class_("SubDevice")
```
I don't like having to define the SubDevice class twice (here and in config.py) but can't find a nice way to avoid circular dependencies.
Could work if it was defined "publicly" in this file and imported to core.py, but kind of think it don't belong to `config_validation.py`.
An alternative could be to move declaration to `core/__init__.py` but that seem a bit drastic.

dala318 replied:
CoPilot proposed using lazy imports so will give that an attempt (and see if the clang-tidy etc. agrees to doing it so)

dala318 committed:
More namespace and import fixes

dala318 committed:
Attempt fixing circular import by lazy import

dala318 committed:
Merge branch 'dev' into multi_device

dala318 committed:
Fix proto generation and clang

bdraco commented on esphome/core/config.py:
```diff
@@ -405,3 +416,12 @@ async def to_code(config):
+    if config[CONF_SUB_DEVICES]:
+        for dev_conf in config[CONF_SUB_DEVICES]:
+            dev = cg.new_Pvariable(dev_conf[CONF_ID])
+            cg.add(dev.set_uid(hash(str(dev_conf[CONF_ID])) % 0xFFFFFFFF))
```
The UID generation using hash(str(device)) % 0xFFFFFFFF could cause collisions

dala318 replied:
Do you have any proposals?
I can concatenate name as well to add more variation.
Could generate just some random id, but thinking to use device id was to use something that generates the same uid every compilation (unless you change id of course)

bdraco replied:
we use fnv1a_32 for cases like this as the collision risk is much lower

bdraco replied:
esphome/core/helpers.h:uint32_t fnv1_hash(const std::string &str); is likely good enough for this

bdraco replied:
But this is happening in python ... hmm

bdraco replied:
https://github.com/Bluetooth-Devices/fnv-hash-fast/blob/main/src/fnv_hash_fast/fnv_wrapper.h

bdraco replied:
`hash()` also isn't stable across python versions

bdraco replied:
```diff
diff --git a/esphome/core/config.py b/esphome/core/config.py
index f3d8b7e71..f4a31c535 100644
--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -38,6 +38,15 @@ from esphome.helpers import copy_file_if_changed, get_str_env, walk_files
 
 _LOGGER = logging.getLogger(__name__)
 
+
+def fnv1a_32bit_hash(string: str) -> int:
+    """FNV-1a 32-bit hash function."""
+    hash_value = 2166136261
+    for char in string:
+        hash_value ^= ord(char)
+        hash_value = (hash_value * 16777619) & 0xFFFFFFFF
+    return hash_value
+
 StartupTrigger = cg.esphome_ns.class_(
     "StartupTrigger", cg.Component, automation.Trigger.template()
 )
@@ -420,7 +429,7 @@ async def to_code(config):
     if config[CONF_SUB_DEVICES]:
         for dev_conf in config[CONF_SUB_DEVICES]:
             dev = cg.new_Pvariable(dev_conf[CONF_ID])
-            cg.add(dev.set_uid(hash(str(dev_conf[CONF_ID])) % 0xFFFFFFFF))
+            cg.add(dev.set_uid(fnv1a_32bit_hash(str(dev_conf[CONF_ID]))))
             cg.add(dev.set_name(dev_conf[CONF_NAME]))
             cg.add(dev.set_area(dev_conf[CONF_AREA]))
             cg.add(cg.App.register_sub_device(dev))
```

dala318 replied:
Added function `fnv1a_32bit_hash` to helpers.py since it needs to be used in at least two places.

bdraco commented on esphome/core/config.py:
```diff
@@ -405,3 +416,12 @@ async def to_code(config):
+    if config[CONF_SUB_DEVICES]:
+        for dev_conf in config[CONF_SUB_DEVICES]:
+            dev = cg.new_Pvariable(dev_conf[CONF_ID])
+            cg.add(dev.set_uid(hash(str(dev_conf[CONF_ID])) % 0xFFFFFFFF))
```
How do we validate Sub-device IDs are unique within a configuration

dala318 replied:
Will not the schema validation solve that?

cv.GenerateID(CONF_ID): cv.declare_id(SubDevice),

jesserockz replied:
Yes declare id will ensure uniqueness

bdraco commented on esphome/core/config.py:
```diff
@@ -405,3 +416,12 @@ async def to_code(config):
+            dev = cg.new_Pvariable(dev_conf[CONF_ID])
+            cg.add(dev.set_uid(hash(str(dev_conf[CONF_ID])) % 0xFFFFFFFF))
+            cg.add(dev.set_name(dev_conf[CONF_NAME]))
+            cg.add(dev.set_area(dev_conf[CONF_AREA]))
```
It seems like devices could be in the same area, however they are only set by name

dala318 replied:
As far as I have understood Home Assistant this thing will only be used at device creation the first time to propose an Area for the device. If the user selects another area it will be stored and used thereafter.

bdraco commented on esphome/components/api/api.proto:
```diff
@@ -188,6 +188,12 @@ message DeviceInfoRequest {
+message SubDeviceInfo {
+  uint32 uid = 1;
+  string name = 2;
+  string suggested_area = 3;
```
Area seems like it should have its own id since multiple devices can be in the same area

dala318 replied:
Could be done but in my opinion that would only complicate things. Fairly easy to just duplicate strings, or use substitutions in yaml.

I suppose the sub-devices could inherit area from main config as default if given there.

bdraco replied:
I'm thinking down the road that the ESP might end up being a hub for devices in the future in different areas.  The area implementation in HA may be improved in the future to better deal with overlapping names so it would be good to have a stable id to be able to join multiple areas

nielsnl68 replied:
Yes, @bdraco i like to use this PR idea to extend the futures of my espnow component indeed.

dala318 replied:
So should I add some core component `area` which both the main esphome config and sub_devices config reference? Or is it a sub-section of the esphome config?

bdraco commented on esphome/core/sub_device.h:
```diff
@@ -0,0 +1,22 @@
+ public:
+  void set_uid(uint32_t uid) { uid_ = uid; }
+  uint32_t get_uid() { return uid_; }
+  void set_name(std::string name) { name_ = std::move(name); }
```
I wonder if StringRef makes sense here

dala318 replied:
I did use StringRef for a while in earlier versions but it only made me do more conversions.
If there are benefits of still using it I can change back.

One more reason I changed to std::string was also that when in a component referencing original StringRef definition in Python it created circular dependencies.

bdraco committed:
Merge branch 'dev' into multi_device

nielsnl68 commented:
> > > As long you can make as open as possible to hook other ideas into the same infrastructure i'm already happy.
> > > At a later point we can think of a way to implement something like the bluetooth proxy.
> > 
> > 
> > I'm all for open an expandable implementation. So if you have ideas what might be needed for making this more compatible to your needs please make suggestions.
> 
> For now, please make it bidirectional. And i think everything is fine.

One more suggestion is to have the option to allow adding new sub-devices dynamically when a new ESPNOW device connect's to the receiver device.
For me this is just a future wish not something to have now. But when you can have this in mind while working on this i would be very happy. I have no clue about the possibility of my suggestion.

bdraco committed:
Merge branch 'dev' into multi_device

dala318 commented:
> > > > > As long you can make as open as possible to hook other ideas into the same infrastructure i'm already happy.
> > > > > At a later point we can think of a way to implement something like the bluetooth proxy.
> > > > 
> > > > 
> > > > I'm all for open an expandable implementation. So if you have ideas what might be needed for making this more compatible to your needs please make suggestions.
> > > 
> > > 
> > > For now, please make it bidirectional. And i think everything is fine.
> > 
> > One more suggestion is to have the option to allow adding new sub-devices dynamically when a new ESPNOW device connect's to the receiver device. For me this is just a future wish not something to have now. But when you can have this in mind while working on this i would be very happy. I have no clue about the possibility of my suggestion.
> 
> I think its important to make sure we can do that before moving this forward as it could be very hard to add later without breaking something with the current design.

So should the sub-devices (and others in the future) be kept in a dynamic array?

Next possible issue down the road if so, the sub-devices are currently sent over via the DeviceInfoResponse proto package. Is that sent cyclic?

If thinking reusable design it sound for me more that other connected devices when the ESPHome instance act as gateway should extend on how the BluetoothProxy solves the same.

If it all could be one solution I'm all OK with that but don't have any clear proposal on how to do it or any of the alternatives.

nielsnl68 commented:
> So should the sub-devices (and others in the future) be kept in a dynamic array?
I think that could be the way to go. But check how the BluetoothProxy does it do.

> If thinking reusable design it sound for me more that other connected devices when the ESPHome instance act as gateway should extend on how the BluetoothProxy solves the same.
I am just a side-liner in this and have no idea of it is easy to implement it. That said i think that you should use the same technique indeed.

> If it all could be one solution I'm all OK with that.
I agree with this. 

> but don't have any clear proposal on how to do it or any of the alternatives.
Maybe have a check how the BluetoothProxy it could result in a proposal.

dala318 committed:
Change hash method

dala318 committed:
Move fnv1a_32bit_hash to helpers

dala318 commented:
> > So should the sub-devices (and others in the future) be kept in a dynamic array?
> 
> I think that could be the way to go. But check how the BluetoothProxy does it do.
> 
This is a bit outside my comfort zone to make statements about, but Bluetooth Proxy seem to have its own proto messages for communication with Home Assistant `BluetoothDeviceRequest` and `BluetoothDeviceConnectionResponse`.
But neither of these seem to have either `name` or `suggested_area` as payload.

So for this to work as some type of coordinating device registry between Virtual Sub Devices and Physical Connected Devices (Bluetooth, ESPNow etc.) some other more advanced solution is needed. And I would need some help from others to propose solutions.

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into multi_device

bdraco commented on esphome/core/config.py:
```diff
@@ -417,3 +433,12 @@ async def to_code(config):
+            cg.add(dev.set_uid(fnv1a_32bit_hash(str(dev_conf[CONF_ID]))))
+            cg.add(dev.set_name(dev_conf[CONF_NAME]))
+            cg.add(dev.set_area(dev_conf[CONF_AREA]))
+            cg.add(cg.App.register_sub_device(dev))
```
We could pre-reserve this vector for ones that are known at build time

bdraco replied:
done

bdraco committed:
cleanups to address review comments

bdraco committed:
cleanups to address review comments

bdraco committed:
cleanups to address review comments

bdraco committed:
cleanups to address review comments

bdraco commented on esphome/core/sub_area.h:
```diff
@@ -0,0 +1,20 @@
+ public:
+  void set_area_id(uint32_t area_id) { area_id_ = area_id; }
+  uint32_t get_area_id() { return area_id_; }
+  void set_name(std::string name) { name_ = std::move(name); }
```
We might be able to get way with const char * here if we always define these at build time and save a few bytes of RAM

bdraco replied:
done

bdraco commented on esphome/core/sub_device.h:
```diff
@@ -0,0 +1,22 @@
+ public:
+  void set_device_id(uint32_t device_id) { device_id_ = device_id; }
+  uint32_t get_device_id() { return device_id_; }
+  void set_name(std::string name) { name_ = std::move(name); }
```
We might be able to get way with const char * here if we always define these at build time and save a few bytes of RAM

bdraco replied:
done

bdraco commented:
Also I was able to significantly reduce the changes in `api_connection.*`

bdraco committed:
fixes

bdraco committed:
revert

dala318 commented on tests/components/esphome/common.yaml:
```diff
@@ -17,10 +17,13 @@ esphome:
     version: "1.1"
     on_update:
       logger.log: on_update
+  sub_areas:
+    - id: another_area
+      name: Another area
```
If going this path, should then not also the normal `area` setting above use the same logic?

dala318 replied:
I do understand and agree to this approach way of keeping data structured and avoid sending possible identical strings several times, but config wise it not as intuitive or requires several lines of config to just get a string passed to backend.
At least propose to rename `sub_areas` to `areas` or `sub_device_areas`, sub-area implies it a part of some other area.

bdraco replied:
If we change the existing suggested area we end up with a breaking change...

bdraco replied:
I guess we could rename `SubAreaInfo` have a top level `AreaInfo`  and migrate over time so we get rid of `suggested_area` eventually but leave if for backwards compat.

bdraco replied:
I think that works. Let me implement that

bdraco replied:
Implemented that. Made it DeviceInfo to be consistent

bdraco committed:
fixes

bdraco commented:
added missing device_id to ListEntitiesSirenResponse which should fix the error

bdraco committed:
migrate to using same area info for top level and sub devices

bdraco committed:
migrate to using same area info for top level and sub devices

bdraco committed:
migrate to using same area info for top level and sub devices

bdraco committed:
migrate to using same area info for top level and sub devices

bdraco committed:
migrate to using same area info for top level and sub devices

bdraco committed:
make areas and devices consistant

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into multi_device

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
make areas and devices consistant

bdraco committed:
reverse space in vectors

bdraco committed:
reverse space in vectors

bdraco commented:
@nielsnl68 I think you should be able to handle this in your ESPNOW PR by defining a `device_id` for each peer which maps to the configured device here

bdraco commented:
I'm pretty happy with this now. I'd like to get a 👍 from another reviewer on the protobuf structure before we update aioesphomeapi.  Once we have that, I can take care of getting a release done for that and than we can write some integration tests to make sure the API works how we want it.

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
preen

bdraco commented:
Will try to look again tomorrow as it's getting late here and have another flight tomorrow

nielsnl68 commented:
... and we know now what we need to do to make the changes to allow late connects. 
That is everything i was hopping for.

bdraco commented on esphome/core/config.py:
```diff
@@ -417,3 +455,70 @@ async def to_code(config):
+                area_conf,
+            )
+            # Create a synthetic area for backwards compatibility
+            area_var = cg.Pvariable(f"area_{area_slug}", Area)
```
We can just use the slug here

bdraco commented on esphome/core/config.py:
```diff
@@ -417,3 +455,70 @@ async def to_code(config):
+            )
+            # Create a synthetic area for backwards compatibility
+            area_var = cg.Pvariable(f"area_{area_slug}", Area)
+            area_id = fnv1a_32bit_hash(area_conf)
```
We need to raise on hash collision. It should be extremely rare but will be very confusing if it does happen

bdraco commented on esphome/core/config.py:
```diff
@@ -417,3 +455,70 @@ async def to_code(config):
+        if areas := config[CONF_AREAS]:
+            for area_conf in areas:
+                area = cg.new_Pvariable(area_conf[CONF_ID])
+                area_id = fnv1a_32bit_hash(str(area_conf[CONF_ID]))
```
We need to raise on hash collision. It should be extremely rare but will be very confusing if it does happen

bdraco commented on esphome/core/config.py:
```diff
@@ -417,3 +455,70 @@ async def to_code(config):
+        # Process devices
+        for dev_conf in devices:
+            dev = cg.new_Pvariable(dev_conf[CONF_ID])
+            cg.add(dev.set_device_id(fnv1a_32bit_hash(str(dev_conf[CONF_ID]))))
```
We need to raise on hash collision. It should be extremely rare but will be very confusing if it does happen

bdraco committed:
handle collisions

bdraco committed:
handle collisions

bdraco committed:
handle collisions

bdraco committed:
handle collisions

bdraco committed:
fixes

bdraco commented:
bump https://github.com/esphome/esphome/pull/9173

bdraco committed:
fixes

bdraco committed:
Merge branch 'dev' into multi_device

bdraco commented:
We are going to need end to end integration tests for this change as well once we have aioesphomeapi updated

bdraco committed:
Merge remote-tracking branch 'dala318/multi_device' into multi_device

bdraco committed:
Merge branch 'dev' into multi_device

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
legacy test

bdraco committed:
one more place to check

bdraco committed:
preen

bdraco committed:
preen

bdraco committed:
simplify

bdraco committed:
simplify

bdraco committed:
simplify

bdraco committed:
simplify

bdraco committed:
units

bdraco committed:
add files

bdraco committed:
adjust

bdraco committed:
cleanups

bdraco commented:
I added all the tests for this as well.  It should be ready for another reviewer at this point.

bdraco committed:
validate sooner

bdraco committed:
validate sooner

bdraco committed:
validate sooner

bdraco committed:
validate sooner

bdraco committed:
validate sooner

bdraco committed:
validation should happen sooner

bdraco committed:
document design decisions

bdraco committed:
lint

bdraco committed:
lint

bdraco committed:
lint

bdraco committed:
lint

bdraco committed:
cleanups

bdraco committed:
Merge branch 'dev' into multi_device

bdraco committed:
remove unreachable code

bdraco committed:
Merge remote-tracking branch 'dala318/multi_device' into multi_device

jesserockz committed:
Dont need to warning about simple string area

A single device in a single area can have a simple string as the area

jesserockz committed:
``this->``

jesserockz committed:
Update testing

bdraco committed:
Merge branch 'dev' into multi_device

bdraco commented:
cd /config
curl -o- -sSL https://gist.githubusercontent.com/bdraco/43f8043cb04b9838383fd71353e99b18/raw/core_integration_pr | bash /dev/stdin -d esphome -p 147343


You can install HA PR as a custom component and restart. It might work on stable HA if there aren't too many changes. You'll also need to copy translations/XXX.json to /config/custom_components/esphome/translations from /usr/src/homeassistant/components/esphome/translations/

jesserockz committed:
Merge branch 'dev' into multi_device

jesserockz committed:
Store reference to device on EntityBase

This is so we can get the name of the device to use as part of the object id and to internally set the name for logging.

bdraco commented on esphome/core/entity_base.cpp:
```diff
@@ -29,16 +36,21 @@ void EntityBase::set_icon(const char *icon) { this->icon_c_str_ = icon; }
+  std::string suffix = "";
+#ifdef USE_DEVICES
+  if (this->device_ != nullptr) {
+    suffix = "@" + str_sanitize(str_snake_case(this->device_->get_name()));
```
I think we need to use device_id here so we known if its still valid or not and we have a reference point

bdraco replied:
I'm not sure this is going to work.  It gets super messy to move the entities between devices
```diff
diff --git a/homeassistant/components/esphome/entity.py b/homeassistant/components/esphome/entity.py
index fcf1260de39..af43d5dae4f 100644
--- a/homeassistant/components/esphome/entity.py
+++ b/homeassistant/components/esphome/entity.py
@@ -68,17 +68,62 @@ def async_static_info_updated(
 
         # Create new entity if it doesn't exist
         if not (old_info := current_infos.pop(info.key, None)):
-            entity = entity_type(entry_data, platform.domain, info, state_type)
-            add_entities.append(entity)
+            # Check if this entity might have moved from another device (object_id changed)
+            # Try to find an entity with a different object_id but same base properties
+            new_unique_id = build_unique_id(device_info.mac_address, info)
+            
+            # Look for entities with different object_id patterns (e.g., temperature@old_device)
+            # that might have moved to this device
+            found_migration = False
+            if "@" in info.object_id:
+                # Extract base object_id without device suffix
+                base_object_id = info.object_id.split("@")[0]
+                
+                # Search through all entities of this platform to find potential matches
+                for entity_entry in ent_reg.entities.get_entries_for_domain(platform.domain):
+                    if (entity_entry.platform == DOMAIN and 
+                        entity_entry.unique_id and 
+                        entity_entry.unique_id != new_unique_id):
+                        # Check if this could be the same entity that moved devices
+                        # This is a heuristic - we look for entities with similar patterns
+                        try:
+                            # Parse the unique_id to check if it's the same base entity
+                            # that just moved to a different device
+                            old_entity_id = entity_entry.entity_id
+                            if base_object_id in old_entity_id:
+                                # Found a potential match - migrate it
+                                ent_reg.async_update_entity(
+                                    entity_entry.entity_id,
+                                    new_unique_id=new_unique_id
+                                )
+                                found_migration = True
+                                break
+                        except Exception:
+                            continue
+            
+            if not found_migration:
+                # No migration found, create new entity
+                entity = entity_type(entry_data, platform.domain, info, state_type)
+                add_entities.append(entity)
             continue
 
-        # Entity exists - check if device_id has changed
+        # Entity exists - check if device_id or object_id has changed
+        old_unique_id = build_unique_id(device_info.mac_address, old_info)
+        new_unique_id = build_unique_id(device_info.mac_address, info)
+        
+        if old_unique_id != new_unique_id:
+            # Unique ID changed (likely due to object_id change when moving devices)
+            entity_id = ent_reg.async_get_entity_id(platform.domain, DOMAIN, old_unique_id)
+            if entity_id:
+                # Update the entity's unique_id to the new one
+                ent_reg.async_update_entity(entity_id, new_unique_id=new_unique_id)
+        
+        # Check if device assignment changed
         if old_info.device_id == info.device_id:
             continue
 
         # Entity has switched devices, update its device assignment
-        unique_id = build_unique_id(device_info.mac_address, info)
-        entity_id = ent_reg.async_get_entity_id(platform.domain, DOMAIN, unique_id)
+        entity_id = ent_reg.async_get_entity_id(platform.domain, DOMAIN, new_unique_id)
         # entity_id should never be None here because old_info is not None,
         # which means the entity was previously created and is in the registry
         assert entity_id is not None
```

bdraco replied:
So I think we don't need to add it to the object id because we already have the device_id

bdraco committed:
fix conflicts

bdraco committed:
fixes

bdraco commented on esphome/core/entity_base.cpp:
```diff
@@ -33,8 +40,7 @@ std::string EntityBase::get_object_id() const {
     return str_sanitize(str_snake_case(App.get_friendly_name()));
-  } else {
-    // `App.get_friendly_name()` is constant.
+  } else {  // `App.get_friendly_name()` is constant.
```
```suggestion
  } else {
    // `App.get_friendly_name()` is constant.
```

bdraco committed:
Update esphome/core/entity_base.cpp

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco commented on esphome/core/application.h:
```diff
@@ -102,11 +109,18 @@ class Application {
       this->friendly_name_ = friendly_name;
     }
-    this->area_ = area;
+    // area is now handled through the areas system
```
```suggestion
```

bdraco committed:
Update esphome/core/application.h

bdraco committed:
cleanup

bdraco committed:
preen

bdraco committed:
cleanup

bdraco committed:
raise

bdraco committed:
migrate

bdraco committed:
migrate

bdraco committed:
migrate

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
preen

bdraco committed:
fixes

bdraco committed:
fixes

bdraco commented:
now validates duplicates
```
esphome compile ol.yaml 
INFO ESPHome 2025.7.0-dev
INFO Reading configuration ol.yaml...
INFO Detected timezone 'Europe/Madrid'
Failed config

sensor.template: [source ol.yaml:719]
  
  Duplicate sensor entity with name 'Lambda Sensor 199' found. Each entity on a device must have a unique name within its platform.
  platform: template
  name: Lambda Sensor 199
  lambda: return 199.0;
```

bdraco commented:
## Breaking Changes

### Duplicate Entity Names Now Fail Validation

Previously, ESPHome would silently accept duplicate entity names within the same platform and device. This caused problems in Home Assistant where the second entity would not appear and generate warnings in the logs about duplicate unique IDs.

**What's changed:**
- Duplicate entity names on the same device and platform now fail validation with a clear error message
- The error tells you exactly which entity is duplicated and where
- This prevents the confusing situation where entities work in ESPHome but fail to appear in Home Assistant

**What you need to do:**
If your configuration has duplicate entity names, you'll see an error like:
```
Duplicate binary_sensor entity with name 'Motion' found. Each entity on a device must have a unique name within its platform.
```

To fix this, give each entity a unique name:
```yaml
# Before (would previously be accepted, now fails):
binary_sensor:
  - platform: gpio
    name: "Motion"
    pin: GPIO5
  - platform: gpio
    name: "Motion"  # ERROR: Duplicate name
    pin: GPIO6

# After (fixed):
binary_sensor:
  - platform: gpio
    name: "Front Motion"
    pin: GPIO5
  - platform: gpio
    name: "Back Motion"
    pin: GPIO6
```

**Note:** This only applies to entities on the same device and platform. You can still have:
- Same names across different platforms (e.g., a `sensor.temperature` and `binary_sensor.temperature`)
- Same names on different devices (using the new `device_id` feature)

bdraco committed:
Merge branch 'dev' into multi_device

bdraco committed:
fix

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
update tests now that duplicate names are validated

bdraco committed:
use a common that does not have dupes on dev

Drafteed commented:
@bdraco @dala318 Thank you, you did a great job! 🎉

I've been waiting for this feature in ESPHome for a long time. Now I can finally separate my 5 virtual devices with a bunch of different platforms on one ESP into separate devices. Thank you 👍. I'll be happy to test this feature during beta testing.

Drafteed commented:
@bdraco After testing i got some things:

**First thing:**

```yaml
esphome:
  name: esphome-hub
  devices:
    - id: ringbell_device
      name: "Ringbell"
    - id: intercom_device
      name: "Intercom"
    - id: le2_device
      name: "Energy Meter"
    - id: tion_device
      name: "TION Breezer"
    - id: xgimi_device
      name: "XGIMI Projector (BLE)"

climate:
  # Expect that device name will be used
  - platform: tion
    id: tion_climate
    device_id: tion_device

fan:
  # Expect that device name will be used
  - platform: tion
    id: tion_fan
    device_id: tion_device
```

After adding the devices to the HA, these two entities will be missing. Adding `name: ""` to each solves this.

**Second thing:**

```yaml
binary_sensor:
  # Ringbell device
  - platform: gpio
    id: ringbell_call_binary
    pin:
      number: GPIO4
      mode:
        input: true
        pulldown: true
    name: "Call"
    device_id: ringbell_device
    device_class: sound

  # Intercom device
  - platform: gpio
    id: intercom_call_binary
    pin:
      number: GPIO5
      inverted: true
      mode:
        input: true
        pullup: true
    filters:
      - delayed_on: 40ms
      - delayed_off: 4000ms
    name: "Call"
    device_id: intercom_device
    device_class: sound
```

Two different virtual devices are using the same name. After adding them to HA, all devices added successfully. `binary_sensor.intercom_call` shows as `unknown`.

Logs:
```
Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 1470, in async_init
    subscription("added", flow.flow_id)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 398, in async_on_flow_init_remove
    "flow": hass.config_entries.flow.async_get(flow_id),
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 234, in async_get
    raise UnknownFlow
homeassistant.data_entry_flow.UnknownFlow

Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1535, in _async_process_registry_update_or_remove
    self.async_write_ha_state()
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1018, in async_write_ha_state
    self._async_write_ha_state()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1140, in _async_write_ha_state
    self.__async_calculate_state()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1080, in __async_calculate_state
    state = self._stringify_state(available)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1024, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 546, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/esphome/sensor.py", line 102, in native_value
    if not self._has_state or (state := self._state).missing_state:
           ^^^^^^^^^^^^^^^
AttributeError: 'EsphomeSensor' object has no attribute '_has_state'. Did you mean: '__attr_state'?
```

After rebooting HA, all sensors (except buttons) become unavailable. No errors in the logs.

When I reboot the ESP device, all sensors are restored, except `binary_sensor.intercom_call`:
```
Platform esphome does not generate unique IDs. ID 9C:9C:1F:C6:49:50-binary_sensor-call@1130295909 is already used by binary_sensor.intercom_call - ignoring binary_sensor.Intercom_Call
```

If I reboot HA again, all sensors will become unavailable again.

To fix this I had to rename the entities in esp and manually cleanup `deleted_entities` from `core.entity_registry` due to https://github.com/home-assistant/core/pull/145278 and https://github.com/home-assistant/core/pull/145191

However, the error when adding a device still exists.

Drafteed commented:
> It compiles good but how did you validate that it works as 2 devices?
> i did expect to see the thermostat in the area livingroom or in Dutch woonkamer

@martijnbuts This only works in beta version of HA 2025.7.0.

martijnbuts commented:
> > It compiles good but how did you validate that it works as 2 devices?
> > i did expect to see the thermostat in the area livingroom or in Dutch woonkamer
> 
> @martijnbuts This only works in beta version of HA 2025.7.0.

i do use the dev build from this morning. I also think that with older versions i did get a compile error. 
i am sure i use the dev build from this morning because i found a duplicate entity i had to remove to be able to create the esp32-32 firmware.
